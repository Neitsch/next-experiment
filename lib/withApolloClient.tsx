import { ApolloClient } from "apollo-client";
/* tslint:disable-next-line */
import Head from "next/head";
import React from "react";
import { getDataFromTree } from "react-apollo";

import initApollo from "./initApollo";

export default App => {
  return class Apollo extends React.Component {
    public static displayName = "withApollo(App)";
    public static async getInitialProps(ctx) {
      const { Component, router } = ctx;
      let appProps = {};
      if (App.getInitialProps) {
        appProps = await App.getInitialProps(ctx);
      }
      const apolloState = {
        data: {},
      };
      const apollo = initApollo({});
      try {
        await getDataFromTree(
          <App
            {...appProps}
            Component={Component}
            router={router}
            apolloState={apolloState}
            apolloClient={apollo}
          />,
        );
      } catch (error) {
        /* tslint:disable-next-line */
        console.error("Error while running `getDataFromTree`", error);
      }

      if (!process.browser) {
        Head.rewind();
      }
      apolloState.data = apollo.cache.extract();

      return {
        ...appProps,
        apolloState,
      };
    }

    private apolloClient: ApolloClient<any>;

    constructor(props) {
      super(props);
      this.apolloClient =
        props.apolloClient || initApollo(props.apolloState.data);
    }

    public render() {
      return <App {...this.props} apolloClient={this.apolloClient} />;
    }
  };
};
