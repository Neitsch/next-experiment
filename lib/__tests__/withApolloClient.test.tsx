import { shallow } from "enzyme";
import React from "react";

import withApolloClient from "../withApolloClient";

jest.mock("apollo-client", () => ({
  ApolloClient: ({ children }) => children,
}));
jest.mock("../initApollo", () => data => ({
  cache: {
    extract: () => ({}),
  },
}));
jest.mock("next/head", () => ({
  rewind: jest.fn(),
}));

/* tslint:disable */
class TestClassWithInit extends React.Component {
  public static async getInitialProps(ctx) {
    TestClass.ctx = ctx;
  }
  private static ctx: any;
  public render() {
    return <div>{TestClass.ctx}</div>;
  }
}
class TestClass extends React.Component {
  private static ctx: any;
  public render() {
    return <div>{TestClass.ctx}</div>;
  }
}
/* tslint:enable */

describe("SSR", () => {
  it("Renders", async () => {
    jest.doMock("react-apollo", () => ({
      getDataFromTree: async () => <div />,
    }));
    const App = withApolloClient(TestClassWithInit);
    await App.getInitialProps({});
    const rendered = shallow(<App apolloState={{ data: {} }} />);
    expect(rendered).toMatchSnapshot();
  });

  it("Data Exception", async () => {
    process.browser = true;
    jest.doMock("react-apollo", () => ({
      getDataFromTree: async () => {
        throw Error();
      },
    }));
    const App = withApolloClient(TestClass);
    await App.getInitialProps({});
    const rendered = shallow(<App apolloState={{ data: {} }} />);
    expect(rendered).toMatchSnapshot();
  });
});
