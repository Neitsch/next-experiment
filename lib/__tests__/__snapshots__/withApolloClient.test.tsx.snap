// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SSR Data Exception 1`] = `
<TestClass
  apolloClient={
    Object {
      "cache": Object {
        "extract": [Function],
      },
    }
  }
  apolloState={
    Object {
      "data": Object {},
    }
  }
/>
`;

exports[`SSR Renders 1`] = `
<TestClassWithInit
  apolloClient={
    Object {
      "cache": Object {
        "extract": [Function],
      },
    }
  }
  apolloState={
    Object {
      "data": Object {},
    }
  }
/>
`;
